apiserver: implement LetsEncrypt support

rogpeppe opened the pull request:
This enables a controller to be contacted via a real
domain name - it will obtain an official HTTPS certificate
from the letsencrypt service.

It's not easy to test this without doing it for real, so to QA:

- bootstrap a controller with api-port=443
- set the IP address associated with a domain you control to the public
IP address of the new controller.
- edit controllers.yaml to change the addresses associated
with the controller to <yourdomain>:443.
- use the controller.

Note that as is, this is open to the DoS vulnerability described
in https://godoc.org/golang.org/x/crypto/acme/autocert#Manager
under the HostPolicy field. We really want to whitelist a single
domain, but that involves adding controller configuration, so we
reserve that for a future PR.

axw commented on provider/ec2/environ.go:
```diff
@@ -1435,6 +1435,11 @@ func (e *environ) setUpGroups(controllerUUID, machineId string, apiPort int) ([]
 			ToPort:    apiPort,
 			SourceIPs: []string{"0.0.0.0/0"},
 		}, {
+			Protocol:  "tcp",
```
If you set api-port=443, then the rule above should suffice...?

rogpeppe replied:
I thought I'd already removed it...

axw commented on state/autocertcache.go:
```diff
@@ -0,0 +1,70 @@
+package state
```
copyright header

rogpeppe replied:
done

axw commented on state/allcollections.go:
```diff
@@ -190,6 +190,12 @@ func allCollections() collectionSchema {
+		// This collection holds information cached by autocert certificate
+		// acquisition.
+		autocertCacheC: {
+			global: true,
```
rawAccess: true

rogpeppe replied:
done (BTW, tangential remark - the "modelUserLastConnectionC" collection below doesn't seem
to belong in this section as it's not global).

rogpeppe replied:
I'm slightly surprised that I didn't get a panic trying to get the WriteCollection without this.

axw commented on state/autocertcache_test.go:
```diff
@@ -0,0 +1,44 @@
+package state_test
```
copyright header

rogpeppe replied:
done

axw commented on state/autocertcache_test.go:
```diff
@@ -0,0 +1,44 @@
+	c.Assert(err, jc.ErrorIsNil)
+	err = cache.Put(ctx, "b", []byte("bval"))
+	c.Assert(err, jc.ErrorIsNil)
+
```
Perhaps also an explicit test that shows what happens when you specify an existing key? I'm assuming it's OK to overwrite a cache entry?

rogpeppe replied:
done

frankban commented on state/autocertcache.go:
```diff
@@ -0,0 +1,70 @@
+}
+
+// Delete implements autocert.Cache.Delete.
+func (cache autocertCache) Delete(ctx context.Context, name string) error {
```
// The operation succeeds if the document is not found?

rogpeppe replied:
That's part of the autocert.Cache.Delete contract, so I don't think we
need to repeat it here.

rogpeppe commented on apiserver/apiserver.go:
```diff
@@ -522,79 +559,72 @@ func (srv *Server) mongoPinger() error {
-		srv:         srv,
-		cert:        cert,
-		certChanged: certChanged,
+// getCertificate returns the certificate appropriate for a client
```
good catch; deleted.

frankban commented on state/autocertcache_test.go:
```diff
@@ -0,0 +1,44 @@
+	c.Assert(err, gc.Equals, autocert.ErrCacheMiss)
+	c.Assert(data, gc.IsNil)
+
+	err = cache.Delete(ctx, "b")
```
Try to delete another time? I kind of feel that these could be different tests, like one for a notFound, one for fetting the cache, one for overriding a name, one for deleting, even twice. Take it or leave it.

rogpeppe replied:
Yeah, I agree - it started off too simple to justify that, but I've now factored it into separate tests.

frankban commented:
This is great Roger, thank you! :+1: with a couple of minor changes.

rogpeppe commented on apiserver/apiserver.go:
```diff
@@ -522,79 +559,72 @@ func (srv *Server) mongoPinger() error {
+func (srv *Server) getCertificate(clientHello *tls.ClientHelloInfo) (*tls.Certificate, error) {
+	logger.Infof("getCertificate - clientHello.ServerName: %q", clientHello.ServerName)
+	cert, shouldUse := srv.localCertificate(clientHello.ServerName)
+	if shouldUse {
```
The idea is that we always fall back to *some* certificate even when we can't obtain a letsencrypt certificate, so you'll get the same failure mode as previously in that case.

jameinel replied:
It might be clearer if we just did:

if !shouldUse {
  logger.Log()
}
return cert, nil

Rather than making it look like we do something different based on shouldUse, just make it clear that all we're doing is adding logging.

rogpeppe commented on apiserver/apiserver.go:
```diff
@@ -146,16 +151,45 @@ func newServer(s *state.State, lis *net.TCPListener, cfg ServerConfig) (_ *Serve
+	}
+	tlsConfig := utils.SecureTLSConfig()
+	tlsConfig.GetCertificate = func(clientHello *tls.ClientHelloInfo) (*tls.Certificate, error) {
+		cert, shouldUse := srv.localCertificate(clientHello.ServerName)
```
done

rogpeppe commented:
> Why do we have to use 443? Are we not allowed to serve valid traffic from a nonstandard port? I realize you probably can't just browse to the web page by the DNSName but you can always go to https://foobar.com:17070 can't you?
> I guess we want to be serving the GUI on :443 so people get it easily?
> 
> I think Andrew's point is valid, that either we just change API port everywhere to default to 443, or we serve only the gui on 443, in which case we don't have to touch API port, we can always expose 443, etc.
> 
> Note that if we do need this change, then we need to touch far more than just AWS, since it will affect all of our clouds.
> 
> Digging into this, it seems the ACME protocol wants to use 443?

Yes, you can't get a certificate unless you listen on 443 (I believe it's because 443 is a "privileged port" and it's not uncommon to be able to listen on other ports without having control of the machine).

BTW I couldn't reply directly to this comment because I had deleted the code in question.

> Have we tested how all of this works if you bootstrap multiple times, and have different servers in between? I don't have a great understanding of how Let's Encrypt keeps things up to date. Does it allow as many signed certificates as you like, as long as the DNS record currently points to the record you are trying to create?
> https://letsencrypt.org/docs/rate-limits/ says
> The main limit is Certificates per Registered Domain (20 per week).
> 
> That seems fine for a real site, slightly worried about testing.

I've added a TODO to make it possible to use a different URL for obtaining the certificate. This would allow use of the letsencrypt staging URL as well as potentially other cert providers should they emerge.

> In fact, the only places I could find that call 'stopRenew' are in the autocert test suite itself, which means that any tests we have that might trigger a manager to run, will end up with renewal goroutines that we can't trigger to stop (at least that I can find).

Yes, this appears to be the case. And even stopRenew doesn't actually wait for any existing renewals to complete. I don't think it's a problem for us in practice though, as even if we trigger a manager to run, it won't be able to obtain a certificate, let alone need to renew one. I'll see if I can get some synchronous-stop changes approved upstream in the autocert repo.

> Do we know the lag times on setting up a new host, and updating your DNS records and the time it will take for Letsencrypt to try and find you? I suppose LE could be nearly synchronous (when we set up the server, we immediately ask them to contact us back, and then they do a DNS lookup and then try to contact us there.)

It seems to be pretty much immediate the first time we ask for the certificate.

jameinel commented:
> It seems to be pretty much immediate the first time we ask for the certificate

I would guess the minimum is still how long it takes for DNS to update. That's probably heavily dependent on your DNS records and who you paid for the DNS, etc. I thought minimums were usually in the hour timeframe, and often the default is 24hrs for an update. But if it is a service that is going to be long running, neither of those is going to be a dealbreaker.

rogpeppe commented:
> I would guess the minimum is still how long it takes for DNS to update. That's probably heavily dependent on your DNS records and who you paid for the DNS, etc. I thought minimums were usually in the hour timeframe, and often the default is 24hrs for an update. But if it is a service that is going to be long running, neither of those is going to be a dealbreaker.

No, it's independent of DNS - the DNS needs to have been updated before you use letsencrypt. (the DNS records don't contain certificates) Letsencrypt is just verifying that it can talk to the service that the DNS record corresponds to.

rogpeppe commented:
$$merge$$

rogpeppe commented:
$$merge$$

rogpeppe commented:
$$merge$$
